src/core/Value.h

```diff
@@ -461,3 +464,36 @@ std::ostream& operator<<(std::ostream& stream, const Value::ObjectType& u);
 using VectorType = Value::VectorType;
 using EmbeddedVectorType = Value::EmbeddedVectorType;
 using ObjectType = Value::ObjectType;
+
+class VirtualValueIterator
+{
+public:
+  // These need to know how the particular iterator works, and so must be implemented in
+  // a subclass.
+  virtual void inc();
+  virtual const Value& operator*() const;
+  virtual bool operator==(const VirtualValueIterator& other) const;
+  virtual bool operator!=(const VirtualValueIterator& other) const;
+  // This one can be generically built on top of the others.
+  const Value *operator->() const { return &**this; }
+};
+
+class ValueIterator {
+private:
+  std::shared_ptr<VirtualValueIterator> vi;
+public:
+  ValueIterator(std::shared_ptr<VirtualValueIterator> i) : vi(i) { };
+
+  ValueIterator& operator++() { vi->inc(); return *this; }
+  const Value& operator*() const { return (**vi); }
+  bool operator==(const ValueIterator& other) const { return *vi == *other.vi; }
+  bool operator!=(const ValueIterator& other) const { return *vi != *other.vi; }
+  const Value *operator->() const { return &**this; }
+};
+
+class Values
```

Comment:
Why would you need `Values`, instead of just returning a `std::pair<ValueIterator, ValueIterator>` from `embeddedValues`?

Reply:
Want the caller to be able to say
```
for (v : something) {
}
```
which implies returning an object that has `begin()` and `end()`.

And:
* as best I can tell it's got to be a concrete "leaf" class, because the `for()` wants to implicitly assign it to an invisible iterator variable.
* Don't want to just return the aggregate class itself (VectorType, ObjectType) because there may be more than one way to iterate across the aggregate.  For ObjectType you might want to iterate across keys, or across values, or across {key, value} pairs.